### PR TITLE
feat: add `svelte/html-self-closing`

### PR DESCRIPTION
**Pull request by marekvospel**
Moved to separate PR from ota-meshi/eslint-plugin-svelte#186

## `svelte/html-self-closing`
A simple rule which automatically closes html tags.
```html
<div></div>
<!-- gets fixed to -->
<div>
```
`vue/html-self-closing` also supports custom setting for svg and math elements, which I couldn't simply extract from Svelte AST, so I didn't include it. It can be simply added by checking whether element name is known svg / math element (just like it's done with void elements)

## Coverage
100% test coverage.

~~is yet to be resolved~~
solved

### CODE REVIEW

File: docs/rules/html-self-closing.md
Comment (marekvospel) [inline docs/rules/html-self-closing.md:68]: > `<svelte:self></svelte:self>` seems to be affected by the `normal` option, but I think they need to share the `component` option or have different option.

https://github.com/ota-meshi/eslint-plugin-svelte/pull/186#discussion_r926286056

**Comment by JounQin** [comment]
Sorry, I accidentally changed the review status. 😅

File: src/utils/template-utils.ts
Comment (ota-meshi) [inline src/utils/template-utils.ts:8]: We must also see the `object` field.

```svelte
<script>
import * as Foo from './my-components'
</script>

<Foo.input />
```

File: src/utils/template-utils.ts
Comment (ota-meshi) [inline src/utils/template-utils.ts:16]: I don't think this function is needed. We can use `node.kind === "component"` directly. Type guards are also effective.

File: docs/rules/html-self-closing.md
Comment (marekvospel) [reply docs/rules/html-self-closing.md:68]: @ota-meshi I'm having problem with implementing this. I've made all necessary changes, but with the selector `SvelteElement` the rule logic isn't being fired for some reason, but in the AST playground the `<svelte:head>` element has type`SvelteElement`
image

Comment (ota-meshi) [reply docs/rules/html-self-closing.md:68]: I think we can use `node.kind==="special"` to determine that.

Comment (marekvospel) [reply docs/rules/html-self-closing.md:68]: > I think we can use `node.kind==="special"` to determine that.

I did add that, but I had problems with the tests, now it's fixed

File: src/rules/html-self-closing.ts
Comment (ota-meshi) [inline src/rules/html-self-closing.ts:64]: Currently we only have the `html` option, so we don't need the rest element.

```suggestion
      html: {
        void: "always",
        normal: "always",
        component: "always",
        svelte: "always",
        ...ctx.options?.[0]?.html,
      },
```

File: src/utils/template-utils.ts
Comment (ota-meshi) [inline src/utils/template-utils.ts:1]: I prefer to add it to `src/utils/ast-utils.ts` rather than adding a new file. Could you move the util functions?

File: docs/rules/html-self-closing.md
Comment (ota-meshi) [inline docs/rules/html-self-closing.md:63]: I thought about this, but I think we can remove `html`. We probably don't plan to add options like `svg`. Also, I don't think there are any users who want to set them individually. What do you think?

```suggestion
      "void": "always", // or "always" or "ignore"
      "normal": "always", // or "never" or "ignore"
      "component": "always", // or "never" or "ignore"
      "svelte": "always" // or "never" or "ignore"
```

**Review by ota-meshi — CHANGES_REQUESTED** [review]

File: docs/rules/html-self-closing.md
Comment (marekvospel) [reply docs/rules/html-self-closing.md:63]: Yeah, I don't know about anybody who would set that separately.

Comment (ota-meshi) [reply docs/rules/html-self-closing.md:63]: Could you change it to omit the `html` field?

Comment (marekvospel) [reply docs/rules/html-self-closing.md:63]: yes, I didn't have time during the last 2 days. I'll change it now

**Review by ota-meshi — APPROVED** [review]
LGTM! Thank you!

**Review by JounQin — CHANGES_REQUESTED** [review]

File: docs/rules/html-self-closing.md
Comment (ota-meshi) [reply docs/rules/html-self-closing.md:33]: > `div` is not self-closing in HTML
> 'html-compliant' should be default IMO.

I think we should make the default of `prettier-plugin-svelte` the default behavior rather than the HTML standard. So I think `all` should be the default.
`prettier-plugin-svelte` has long been the default style for the community.
Turn off these rules when users use plugins with `prettier-plugin-svelte`, but I think it makes sense to adopt the same style as `prettier-plugin-svelte` by default.

Comment (ota-meshi) [reply docs/rules/html-self-closing.md:33]: >  we should have three options: 'all' | 'html-compliant' | 'none'

I think the current options can already do the same. Why do you think we need three options?

> ```jsonc
> {
>   "svelte/html-self-closing": [
>     "error",
>     {
>       "void": "always", // or "always" or "ignore"
>       "normal": "always", // or "never" or "ignore"
>       "component": "always", // or "never" or "ignore"
>       "svelte": "always" // or "never" or "ignore"
>     }
>   ]
> }
> ```
> 
> - `void` (`"always"` by default)... Style of HTML void elements
> - `component` (`"always"` by default)... Style of svelte components
> - `svelte` (`"always"` by default)... Style of svelte special elements (`<svelte:head>`, `<svelte:self>`)
> - `normal` (`"always"` by default)... Style of other elements
> 
> Every option can be set to
> - "always" (`<div />`)
> - "never" (`<div></div>`)
> - "ignore" (either `<div />` or `<div></div>`)

Comment (JounQin) [reply docs/rules/html-self-closing.md:33]: @ota-meshi See also https://github.com/sveltejs/prettier-plugin-svelte/issues/284

Comment (ota-meshi) [reply docs/rules/html-self-closing.md:33]: Oh. I didn't know that.
Hmm. The maintainers don't seem to be involved in the discussion, so I'm still not sure if they're thinking of doing so by default.

Comment (JounQin) [reply docs/rules/html-self-closing.md:33]: For example, I'd prefer `<br />` and `<div></div>`, it seems we can't achieve this for current options?

Comment (JounQin) [reply docs/rules/html-self-closing.md:33]: > I think the current options can already do the same. Why do you think we need three options?

Hah, right, it makes sense. Although I still think one option with different values is much easiler to understand and use like prettier https://prettier.io/docs/en/options.html#html-whitespace-sensitivity

Comment (ota-meshi) [reply docs/rules/html-self-closing.md:33]: It's nice to have easier options, but Svelte isn't HTML, so you need to be clear about how each option works.

Comment (dummdidumm) [reply docs/rules/html-self-closing.md:33]: The linked option is not related to tag closing, it's related to whitespace and prettier-plugin-svelte already adheres to that option.

Comment (JounQin) [reply docs/rules/html-self-closing.md:33]: > The linked option is not related to tag closing, it's related to whitespace and prettier-plugin-svelte already adheres to that option.

I know, I'm just saying we can have such `single` option maybe.

Comment (marekvospel) [reply docs/rules/html-self-closing.md:33]: > I know, I'm just saying we can have such `single` option, maybe.

That's a good idea, but I'd also keep the current configuration in overrides section, because there could be someone who only wants self closing on tags in specific category, and it would be waste to just throw away the option to set that.

Comment (ota-meshi) [reply docs/rules/html-self-closing.md:33]: @marekvospel Do you currently have time to work on this PR?

I think the new single option is that we can add features later, so if you don't have time now, after merging this PR, we can work on a new PR when you/we have time.

Comment (marekvospel) [reply docs/rules/html-self-closing.md:33]: @ota-meshi no, I'll have time in 2 weeks or later. I'll create a new PR when I have time

**Pull request merged**
